### ISSUE TITLE
Better Unicode versions support

### ISSUE THREAD

**Issue by jerch** [xtermjs/xterm.js#2668]
Tracking issue for better Unicode version support.

Things to do:
- create a build tool based on https://github.com/fluffos/fluffos/tree/master/src/thirdparty/widecharwidth to extract needed Unicode information for a specific version
- automate version addon creation as far as possible with the build tool

better Unicode support in general:
- establish a setting for ambiguous chars rendered as wide chars
- bidi and grapheme support
- long run: conceptual work towards a better unicode handling in terminals via terminal-wg

**Comment by thefallentree** [comment]
just reminder that we also need have an options to  treat ambiguous character to double width

**Comment by jerch** [comment]
@thefallentree Woops, that was the initial reason to create the issue, thx for pointing out, fixed it.

**Comment by gustavopch** [comment]
Does this issue address overlapping emoji issues in xterm terminals like the one in the following screenshot taken from VSCode integrated terminal?
![image](https://user-images.githubusercontent.com/10941958/109856918-d3e5d600-7c38-11eb-8c88-2687b6cb94c9.png)

OS: Arch Linux 5.11.2-arch1-1
VSCode: 1.53.2
Emoji font: Noto Color Emoji (`noto-fonts-emoji` package from AUR)

**Comment by jerch** [comment]
@gustavopch Yeah ideally most of these issue would be handled by that. Still there are corner cases where a proper result cannot be determined - if app side and the terminal host system have different ideas about a certain codepoint width we have to prioritize one side. Here we would have to go with app side by default to not get out of sync with line length assumption of the app.

This is mostly related to different unicode versions being used on either side and currently there is simply no way to harmonize app and terminal side in this regard. Not sure if we will ever see a good solution to that (would need a really advanced unicode terminal interface), also note that most local terminals do not suffer as much as xterm.js does (due its decoupled terminalview <--> PTY host nature).

**Comment by PerBothner** [comment]
If someone wants to implement extended grapheme clusters, you might find my [unicode-properties library](https://github.com/PerBothner/unicode-properties) useful. In a single efficient lookup it can return both character widths (single, double, or ambiguous) and character classes (for determining grapheme cluster bounderies). This is used by [DomTerm](https://domterm.org) - see [second screenshot on this page](https://domterm.org/Features.html). It would not be difficult to extend unicode-properties to report additional character attributes.

Once you determine cluster boundaries, it should not be so difficult to draw them if using either a dom-based or canvas-based renderer, since you can delegate to the browser's text rendering.

I do think ultimately terminals need to support variable-width fonts, but this is a different and complicated discussion. (I've been pondering this for a while.)
